### DISCUSSION TITLE
How to retrieve a list of supported locales in Next.js 13?

### DISCUSSION THREAD

**Discussion by albertothedev** [amannn/next-intl#263] [Q&A] [answered]
In the Next.js 13 example, the [LocaleSwitcher](https://github.com/amannn/next-intl/blob/main/packages/example-next-13/src/components/LocaleSwitcher.tsx) component toggles between two locally-defined locales. Needing to define the locales in place is bad practice, and this example doesn't work with more than two locales. Just as `useLocale()` retrieves the current locale, is there a way to get the list of supported locales defined in the middleware?

**Comment by amannn** [comment, marked as answer]
The Next.js 13 example is simplistic by intention. You can of course move your configuration to a separate module that you reference in all relevant places.

E.g.:

```ts
// src/config.ts
export default {
  locales: ['en', 'de'],
  defaultLocale: 'en'
};
```

Hope this helps!

**Reply by albertothedev** [reply]
That is similar to what I have ended up doing in my project. One problem with your implementation is that it's not type safe. You could do:

```typescript
// src/config.ts
export default {
  bababooie: true,
};
```

```typescript
// src/middleware.ts
export default createMiddleware(config);
```

What I have done is assign `config`'s type to that of `createMiddleware`'s parameters, which can be done in two ways:

```typescript
// src/config.ts
import MiddlewareConfig from "next-intl/dist/src/middleware/NextIntlMiddlewareConfig";

const config: MiddlewareConfig = { ... };

export default config;
```

or

```typescript
// src/config.ts
import createMiddleware from "next-intl/middleware";

type CreateMiddlewareParameters = Parameters<typeof createMiddleware>[0];

const config: CreateMiddlewareParameters = { ... };

export default config;
```

I have chosen the second implementation because it's safer and easier to understand.

Going back to the example, I think having more than two locales is a sufficiently common use case and the implementation sufficiently involved that it should be featured in the docs and the example code. If you're open to the idea of adding this, I would be willing to open a PR and discuss it further. What do you think?

**Reply by amannn** [reply]
Hmm, for me this is a type error:

<img width="1341" alt="Screenshot 2023-05-02 at 10 53 37" src="https://user-images.githubusercontent.com/4038316/235622882-d8219ed1-2ff2-4148-a971-f6a43de2885e.png">

I'm using `typescript@4.6.3` on the main branch of this repo. Which version are you using?

**Reply by albertothedev** [reply]
My bad, I made a mistake when providing the example for the typing behavior. The issue is that the type checking works differently depending on whether it's an object or a variable containing an object that is passed to `createMidleware`. Here's a showcase of the full behavior:

When passing an object to `createMiddleware`, errors will be thrown when either `locales` or `defaultLocale` are missing and when invalid properties are passed:

- Missing property:
  ![Screenshot_20230502_110828](https://user-images.githubusercontent.com/46131317/235626333-c37fbe29-8a20-4a91-863b-3aafb9bcfa65.png)


- Invalid property:
  ![Screenshot_20230502_111025](https://user-images.githubusercontent.com/46131317/235626714-aea60bf1-1232-46b6-a66d-129f0586264e.png)

Now notice how when it is a variable that is passed to `createMiddleware`, invalid properties don't throw errors:

- Missing property (error, correct behavior):
  ![Screenshot_20230502_111307](https://user-images.githubusercontent.com/46131317/235627594-e745b41a-97d9-401c-9aea-6503bb4496db.png)

- Invalid property (no error, incorrect behavior):
  ![Screenshot_20230502_111430](https://user-images.githubusercontent.com/46131317/235627782-b211024e-9d25-48ce-aa1f-67fdd29153f9.png)

To perfectly replicate the type-checking behavior that we get when directly passing an object to `createMiddleware`, we must assign the proper type to `config`:

- Missing property (error, correct behavior):
  ![Screenshot_20230502_111720](https://user-images.githubusercontent.com/46131317/235628262-5c439d4d-5be6-4d99-86f7-accdb73ab90d.png)

- Invalid property (error, correct behavior):
  ![Screenshot_20230502_111735](https://user-images.githubusercontent.com/46131317/235628463-fb252e52-c6fc-46aa-9189-5877733499a2.png)

**Reply by amannn** [reply]
Thanks for providing the details!

That's correct, this behavior is called [excess property checking](https://observablehq.com/@koop/excess-property-checking-in-typescript). I don't see a practical problem here if you pass additional unknown properties to the middleware, so I think this is ok. You can of course use something like `lodash/pick` if you want to avoid passing unnecessary properties.

**Reply by albertothedev** [reply]
Maybe I am too cautious, but for me the plausible case where I would add a non-existant property thinking of another library's configuration or misspell a property warrant the more strict type checking. When I use TypeScript I want it to behave consistently, like I know it should, instead of ending a day of bug fixing having learned the buzzword for one of its shortcomings.

Regardless of what's done about the type checking, I still think the example should feature more than two locales, which I think is a vastly more common use case. If you're open to the idea, I can submit a PR.

**Reply by amannn** [reply]
Sure, I understand your perspective. There are pros and cons in regard to excess property checking, balancing correctness and pragmatism. I wouldn't want to trade the job with the TypeScript team, it's probably hard enough to keep this balance and there will always be cases where users are put off by their decisions.

Thanks for kindly offering a pull request! I think for the time being I'd prefer the example to focus more on simplicity and the ability to easily copy-paste & modify code, without having to follow file references. I hope you can understand! I'll keep an eye open though for similar inquiries, this was the first time this topic came up so far.

**Reply by albertothedev** [reply]
Sure thing!

Keep up the good work, I have just spent the past weekend looking for an internationalization library for a project using Next.js 13, comparing Next.js 13's built-in support, `next-i18next`, `next-translate`, and this library and have chosen this one because of its features and sensible defaults and the fact that it's the only one that doesn't require you to write your own middleware.

The only missing feature is support for React Server Components but I see it's being worked on. Best of luck with your project!

FYI, in case you haven't heard, the Next.js team has said they will comment on the `app` directory's stability during the Vercel Ship event taking place this week.

**Reply by amannn** [reply]
That's great, glad to hear you've picked `next-intl` and thank you for the kind words!

Yep, [RSC support is currently in beta](https://next-intl-docs.vercel.app/docs/next-13/server-components). I know that many users already use it for their apps, there's a large [feedback thread in the corresponding PR](https://github.com/amannn/next-intl/pull/149). The piece that's currently missing is SSG support, I really hope that upcoming React/Next.js features will enable this. And yes, I'm curiously following along with announcements! 🙂

Let me know how it goes in case you run into something else! 🙌
